File: Example/Stripe iOS Example (Custom)/AlipayExampleViewController.m
//
//  AlipayExampleViewController.m
//  Stripe iOS Example (Custom)
//
//  Created by Ben Guo on 2/22/17.
//  Copyright © 2017 Stripe. All rights reserved.
//

#import <Stripe/Stripe.h>
#import "AlipayExampleViewController.h"
#import "BrowseExamplesViewController.h"

@interface AlipayExampleViewController ()
@property (nonatomic, weak) UIButton *payButton;
@property (nonatomic, weak) UILabel *waitingLabel;
@property (nonatomic, weak) UIActivityIndicatorView *activityIndicator;
@property (nonatomic) STPSource *source;
@end

@implementation AlipayExampleViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    self.view.backgroundColor = [UIColor whiteColor];
    self.title = @"Alipay";
    if ([self respondsToSelector:@selector(setEdgesForExtendedLayout:)]) {
        self.edgesForExtendedLayout = UIRectEdgeNone;
    }

    UIButton *button = [UIButton buttonWithType:UIButtonTypeSystem];
    [button setTitle:@"Pay with Alipay" forState:UIControlStateNormal];
    [button sizeToFit];
    [button addTarget:self action:@selector(pay) forControlEvents:UIControlEventTouchUpInside];
    self.payButton = button;
    [self.view addSubview:button];

    UILabel *label = [UILabel new];
    label.text = @"Waiting for payment authorization";
    [label sizeToFit];
    label.textColor = [UIColor grayColor];
    label.alpha = 0;
    [self.view addSubview:label];
    self.waitingLabel = label;

    UIActivityIndicatorView *activityIndicator = [[UIActivityIndicatorView alloc] initWithActivityIndicatorStyle:UIActivityIndicatorViewStyleGray];
    activityIndicator.hidesWhenStopped = YES;
    self.activityIndicator = activityIndicator;
    [self.view addSubview:activityIndicator];
}

- (void)viewDidLayoutSubviews {
    [super viewDidLayoutSubviews];
    CGFloat padding = 15;
    CGRect bounds = self.view.bounds;
    self.payButton.center = CGPointMake(CGRectGetMidX(bounds), 100);
    self.activityIndicator.center = CGPointMake(CGRectGetMidX(bounds),
                                                CGRectGetMaxY(self.payButton.frame) + padding*2);
    self.waitingLabel.center = CGPointMake(CGRectGetMidX(bounds),
                                           CGRectGetMaxY(self.activityIndicator.frame) + padding*2);
}

- (void)updateUIForPaymentInProgress:(BOOL)paymentInProgress {
    self.navigationController.navigationBar.userInteractionEnabled = !paymentInProgress;
    self.payButton.enabled = !paymentInProgress;
    [UIView animateWithDuration:0.2 animations:^{
        self.waitingLabel.alpha = paymentInProgress ? 1 : 0;
    }];
    if (paymentInProgress) {
        [self.activityIndicator startAnimating];
    } else {
        [self.activityIndicator stopAnimating];
    }
}

- (NSURLRequest *)connection:(NSURLConnection *)connection willSendRequest:(NSURLRequest *)request redirectResponse:(NSURLResponse *)response {
    if (!request) return request;

    NSURL *url = request.URL;
    if ([url.host containsString:@"alipay.com"] ||
         ([url.host isEqualToString:@"stripe.com"] && [url.path isEqualToString:@"/sources/test_redirect"])) {
        [[UIApplication sharedApplication] openURL:url];
        return nil;
    }

    return request;
}

- (void)completeRedirect {
    [self updateUIForPaymentInProgress:NO];

    switch (self.source.status) {
        case STPSourceStatusChargeable:
        case STPSourceStatusConsumed:
            [self.delegate exampleViewController:self didFinishWithMessage:@"Payment successfully created"];
            break;
        case STPSourceStatusCanceled:
            [self.delegate exampleViewController:self didFinishWithMessage:@"Payment failed"];
            break;
        case STPSourceStatusPending:
        case STPSourceStatusFailed:
        case STPSourceStatusUnknown:
            [self.delegate exampleViewController:self didFinishWithMessage:@"Order received"];
            break;
    }

}

#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wdeprecated"
- (void)pay {
    if (![Stripe defaultPublishableKey]) {
        [self.delegate exampleViewController:self didFinishWithMessage:@"Please set a Stripe Publishable Key in Constants.m"];
        return;
    }
    [self updateUIForPaymentInProgress:YES];
    STPSourceParams *sourceParams = [STPSourceParams new];
    sourceParams.rawTypeString = @"alipay";
    sourceParams.redirect = @{
                              @"return_url": @"payments-example://stripe-redirect"
                              };
    sourceParams.currency = @"jpy";
    sourceParams.amount = @(1099);
    [[STPAPIClient sharedClient] createSourceWithParams:sourceParams completion:^(STPSource *source, NSError *error) {
        if (error) {
            [self.delegate exampleViewController:self didFinishWithError:error];
        } else {
            // In order to use STPRedirectContext, you'll need to set up
            // your app delegate to forward URLs to the Stripe SDK.
            // See `[Stripe handleStripeURLCallback:]`
            self.source = source;
            NSURLRequest *request = [NSURLRequest requestWithURL:source.redirect.url];
            [NSURLConnection connectionWithRequest:request delegate:self];
        }
    }];
}
#pragma clang diagnostic pop

@end

File: Example/Custom Integration/AlipayExampleViewController.swift
//
//  AlipayExampleViewController.swift
//  Custom Integration
//
//  Created by Yuki Tokuhiro on 9/16/19.
//  Copyright © 2019 Stripe. All rights reserved.
//

import UIKit
import Stripe

enum PaymentStatus {
    case success
    case cancelled
    case failed(error: Error?)
}

class AlipayExampleViewController: UIViewController {
    @objc weak var delegate: ExampleViewControllerDelegate?
    var redirectContext: STPRedirectContext?
    var inProgress: Bool = false {
        didSet {
            navigationController?.navigationBar.isUserInteractionEnabled = !inProgress
            payButton.isEnabled = !inProgress
            inProgress ? activityIndicatorView.startAnimating() : activityIndicatorView.stopAnimating()
        }
    }
    
    // UI
    lazy var activityIndicatorView = {
       return UIActivityIndicatorView(style: .gray)
    }()
    lazy var payButton: UIButton = {
        let button = UIButton(type: .roundedRect)
        button.setTitle("Pay with Alipay", for: .normal)
        button.addTarget(self, action: #selector(didTapPayButton), for: .touchUpInside)
        return button
    }()
    
    override func viewDidLoad() {
        super.viewDidLoad()
        view.backgroundColor = .white
        title = "Alipay"
        [payButton, activityIndicatorView].forEach { subview in
            view.addSubview(subview)
            subview.translatesAutoresizingMaskIntoConstraints = false
        }

        let constraints = [
            payButton.centerXAnchor.constraint(equalTo: view.centerXAnchor),
            payButton.centerYAnchor.constraint(equalTo: view.centerYAnchor),
            
            activityIndicatorView.centerXAnchor.constraint(equalTo: view.centerXAnchor),
            activityIndicatorView.centerYAnchor.constraint(equalTo: view.centerYAnchor),
        ]
        NSLayoutConstraint.activate(constraints)
    }
    
    @objc func didTapPayButton() {
        guard Stripe.defaultPublishableKey() != nil else {
            delegate?.exampleViewController(self, didFinishWithMessage: "Please set a Stripe Publishable Key in Constants.m")
            return;
        }
        inProgress = true
        pay()
    }
    
    func didFinishPayment(_ status: PaymentStatus) {
        inProgress = false
        switch status {
        case .success:
            delegate?.exampleViewController(self, didFinishWithMessage: "Your order succeeded!")
        case .cancelled:
            return
        case .failed(let error):
            delegate?.exampleViewController(self, didFinishWithError: error)
        }
    }
}

// MARK: -
extension AlipayExampleViewController {
    @objc func pay() {
        // 1. Create an Alipay Source.
        let sourceParams = STPSourceParams.alipayParams(withAmount: 100,
                                                        currency: "USD",
                                                        returnURL: "payments-example://safepay/")
        STPAPIClient.shared().createSource(with: sourceParams) { source, error in
            guard let source = source else {
                self.didFinishPayment(.failed(error: error))
                return
            }
            // 2. Redirect your customer to Alipay.
            // If the customer has the Alipay app installed, we open it.
            // Otherwise, we open alipay.com.
            self.redirectContext = STPRedirectContext(source: source) { sourceID, clientSecret, error in
                guard let clientSecret = clientSecret else {
                    self.didFinishPayment(.failed(error: error))
                    return
                }
                
                // 3. Poll your backend to show the customer their order status.
                // This step is ommitted in the example, as our backend does not track orders.
                self.delegate?.exampleViewController(self, didFinishWithMessage: "Your order was received and is awaiting payment confirmation.")
                
                // 4. On your backend, use webhooks to charge the Source and fulfill the order
            }
            self.redirectContext?.startRedirectFlow(from: self)
        }
    }
}
